src/herder/Upgrades.cpp

```diff
@@ -1029,4 +1242,201 @@ Upgrades::applyReserveUpgrade(AbstractLedgerTxn& ltx, uint32_t newReserve)
+        CLOG_DEBUG(Herder, "Got bad configUpgradeSet {}: no entries updated",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+
+    if (!std::is_sorted(
+            upgradeSetXDR.updatedEntry.begin(),
+            upgradeSetXDR.updatedEntry.end(),
+            [](ConfigSettingEntry const& a, ConfigSettingEntry const& b) {
+                return a.configSettingID() < b.configSettingID();
+            }))
+    {
+        CLOG_DEBUG(Herder,
+                   "Got bad configUpgradeSet {}: the entries are not ordered",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+    if (std::adjacent_find(
+            upgradeSetXDR.updatedEntry.begin(),
+            upgradeSetXDR.updatedEntry.end(),
+            [](ConfigSettingEntry const& a, ConfigSettingEntry const& b) {
+                return a.configSettingID() == b.configSettingID();
+            }) != upgradeSetXDR.updatedEntry.end())
+    {
+        CLOG_DEBUG(Herder, "Got bad configUpgradeSet {}: duplicate entry",
+                   hexAbbrev(key.contentHash));
+        return false;
+    }
+    return true;
+}
+
+ConfigUpgradeSet const&
+ConfigUpgradeSetFrame::toXDR() const
+{
+    return mConfigUpgradeSet;
+}
+
+ConfigUpgradeSetKey const&
+ConfigUpgradeSetFrame::getKey() const
+{
+    return mKey;
+}
+
+LedgerKey
+ConfigUpgradeSetFrame::getLedgerKey(ConfigUpgradeSetKey const& upgradeKey)
+{
+    SCVal v;
+    v.type(SCV_BYTES);
+    v.bytes().insert(v.bytes().begin(), upgradeKey.contentHash.begin(),
+                     upgradeKey.contentHash.end());
+
+    LedgerKey lk;
+    lk.type(CONTRACT_DATA);
+    lk.contractData().contractID = upgradeKey.contractID;
+    lk.contractData().key = v;
+    return lk;
+}
+
+Hash const&
+ConfigUpgradeSetFrame::getHash() const
+{
+    return mKey.contentHash;
+}
+
+bool
+ConfigUpgradeSetFrame::upgradeNeeded(AbstractLedgerTxn& ltx) const
+{
+    for (auto const& updatedEntry : mConfigUpgradeSet.updatedEntry)
+    {
+        LedgerKey key(LedgerEntryType::CONFIG_SETTING);
+        key.configSetting().configSettingID = updatedEntry.configSettingID();
+        bool isSame =
+            ltx.load(key).current().data.configSetting() == updatedEntry;
+        if (!isSame)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+void
+ConfigUpgradeSetFrame::applyTo(AbstractLedgerTxn& ltx) const
+{
+    for (auto const& updatedEntry : mConfigUpgradeSet.updatedEntry)
+    {
+        LedgerKey key(LedgerEntryType::CONFIG_SETTING);
+        key.configSetting().configSettingID = updatedEntry.configSettingID();
+        ltx.load(key).current().data.configSetting() = updatedEntry;
+    }
+}
+
+bool
+ConfigUpgradeSetFrame::isConsistentWith(
+    ConfigUpgradeSetFrameConstPtr const& scheduledUpgrade) const
+{
+    if (scheduledUpgrade == nullptr)
+    {
+        // We don't have any config upgrades scheduled.
+        return false;
+    }
+    return getHash() == scheduledUpgrade->getHash();
```

Comment:
check must compare both contract ID and hash otherwise votes may disagree on where to find the blob

Reply:
Updated to compare keys